Cannot run on Gradle 7 due to `afterEvaluate`

I'm submitting a…

- [x] bug report
- [ ] feature request
- [ ] other

**Short description of the issue/suggestion:**
Running on gradle 7, which is required if running gradle on Java 16 (which is needed due to jlink failure, described in #111), JavaPackager's tasks fail with "afterEvaluate cannot be called after the project has been evaluated."


**Steps to reproduce the issue/enhancement:**

1. Attempt to package an application using Gradle 7

**What is the expected behavior?**
For it to function.


**What is the current behavior?**
Failure.


**Do you have outputs, screenshots, demos or samples which demonstrate the problem or enhancement?**
This is related to https://github.com/gradle/gradle/pull/10831 and the issue it links.

**What is the motivation / use case for changing the behavior?**



**Please tell us about your environment:**

- JavaPackager version: 1.6.0-SNAPSHOT
- OS version: Windows 10
- JDK version: 16
- Build tool: 
	- [ ] Maven 
	- [X] Gradle 

**Other information** (e.g. related issues, suggestions how to fix, links for us to have context)

In a local version with JavaPackager 1.6.0-SNAPSHOT I have fixed this by changing https://github.com/fvarrui/gradle-launch4j/blob/01b03faed4e028874116bd3f1b6f1110e948b87c/src/main/groovy/edu/sc/seis/launch4j/tasks/DefaultLaunch4jTask.groovy#L41 to 
```java
if (project.getState().getExecuted()) {
```
which should maintain behavior of gradle 6 > as the closure wasn't run if the project was already evaluated. This method is from the first resolved comment on the linked PR, that was rejected as it didn't offer any solutions to the root issue - I have applied it to simply maintain the previous behavior.

Small note: I did try to merge upstream's launch4j but it seems to have diverged to the point where JavaPackager fails to function for windows distributions. I did not spend time looking for the cause of this.

Hi @keastrid!

I've finally decided to remove [my fork](https://github.com/fvarrui/gradle-launch4j) of [gradle-launch4j](https://github.com/TheBoegl/gradle-launch4j) project (which was born due to the difficulties I found when including dependencies that were not directly hosted in Maven Central ... problem that I solved recently). So, now JavaPackager depends on the original project, which is currently in its version 2.5.0. I've published this changes on [issue-112](https://github.com/fvarrui/JavaPackager/tree/issue-112). In this way I reduce the number of projects I have to manage.

I've also managed to reproduce your issue using a [sample project](https://github.com/fvarrui/HelloWorldGradle):

```
C:\Users\fvarrui\GitHub\HelloWorldGradle>gradlew --version

------------------------------------------------------------
Gradle 7.1.1
------------------------------------------------------------

Build time:   2021-07-02 12:16:43 UTC
Revision:     774525a055494e0ece39f522ac7ad17498ce032c

Kotlin:       1.4.31
Groovy:       3.0.7
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          16.0.1 (Eclipse OpenJ9 openj9-0.26.0)
OS:           Windows 10 10.0 amd64

C:\Users\fvarrui\GitHub\HelloWorldGradle>gradlew packageForWindows

> Task :packageForWindows
Using packager io.github.fvarrui.javapackager.packagers.WindowsPackager
Creating app ...
    Initializing packager ...
        PackagerSettings [outputDirectory=C:\Users\fvarrui\GitHub\HelloWorldGradle\build, licenseFile=null, iconFile=null, generateInstaller=false, mainClass=io.github.fvarrui.helloworld.Main, name=HelloWorldGradle, displayName=HelloWorldGradle, version=1.0.0, description=HelloWorldGradle, url=null, administratorRequired=false, organizationName=ACME, organizationUrl=, organizationEmail=, bundleJre=true, customizedJre=true, jrePath=null, jdkPath=C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9, additionalResources=[C:\Users\fvarrui\GitHub\HelloWorldGradle\src\main\resources\info.txt], modules=[], additionalModules=[], platform=windows, envPath=null, vmArgs=[], runnableJar=null, copyDependencies=true, jreDirectoryName=jre, winConfig=WindowsConfig [icoFile=null, headerType=gui, companyName=ACME, copyright=ACME, fileDescription=HelloWorldGradle, fileVersion=1.0.0.0, internalName=HelloWorldGradle, language=null, originalFilename=HelloWorldGradle.exe, productName=HelloWorldGradle, productVersion=1.0.0.0, trademarks=null, txtFileVersion=1.0.0, txtProductVersion=1.0.0, disableDirPage=true, disableProgramGroupPage=true, disableFinishedPage=true, createDesktopIconTask=true, generateSetup=true, generateMsi=true, generateMsm=false, msiUpgradeCode=e24d429d-343a-482e-8c56-09accc73d030, wrapJar=true, setupLanguages={english=compiler:Default.isl, spanish=compiler:Languages\Spanish.isl}, setupMode=installForAllUsers, signing=null, registry=Registry [entries=[RegistryEntry [key=HKCU:MyGradleApp, valueName=greeting, valueType=REG_SZ, valueData=hello]]]], linuxConfig=null, macConfig=null, createTarball=false, createZipball=false, extra={}, useResourcesAsWorkingDir=true, assetsDir=C:\Users\fvarrui\GitHub\HelloWorldGradle\assets, classpath=null, jreMinVersion=null, manifest=Manifest [additionalEntries={}, sections=[]], additionalModulePaths=[C:\Users\fvarrui\GitHub\HelloWorldGradle\src]]
    Packager initialized!

    Creating app structure ...
        Removing folder [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle]
        Old app folder removed C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle
        App folder created: C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle
        Assets folder created: C:\Users\fvarrui\GitHub\HelloWorldGradle\build\assets
    App structure created!

    Resolving resources ...
        License file found: C:\Users\fvarrui\GitHub\HelloWorldGradle\LICENSE
        Icon file resolved: C:\Users\fvarrui\GitHub\HelloWorldGradle\assets\windows\HelloWorldGradle.ico
        Effective additional resources [C:\Users\fvarrui\GitHub\HelloWorldGradle\src\main\resources\info.txt, C:\Users\fvarrui\GitHub\HelloWorldGradle\LICENSE, C:\Users\fvarrui\GitHub\HelloWorldGradle\assets\windows\HelloWorldGradle.ico]
    Resources resolved!

    Copying additional resources
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\src\main\resources\info.txt] to folder [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle]
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\LICENSE] to folder [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle]
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\assets\windows\HelloWorldGradle.ico] to folder [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle]
    All additional resources copied!

    Copying all dependencies ...
    Dependencies copied to C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle\libs!

    Creating runnable JAR...
    Runnable jar created in C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle-1.0.0-runnable.jar!

    Bundling JRE ... with C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9
        Creating customized JRE ...
        Getting required modules ...
            Executing command: cmd.exe /s /c ""C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9\bin\jdeps" -q --multi-release 16 --ignore-missing-deps --print-module-deps --module-path C:\Users\fvarrui\GitHub\HelloWorldGradle\src C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle\libs\*.jar C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle-1.0.0-runnable.jar"
            java.base,java.desktop
        Required modules found: [java.base, java.desktop]

        Creating JRE with next modules included: java.base,java.desktop
        Using C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9\jmods modules directory
        Executing command: cmd.exe /s /c ""C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9\bin\jlink" --module-path "C:\Program Files\AdoptOpenJDK\jdk-16.0.1.9-openj9\jmods" --add-modules java.base,java.desktop --output C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle\jre --no-header-files --no-man-pages --strip-debug --compress=2"
        Removing folder [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle\jre\legal]
    JRE bundled in C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle\jre!

    Creating windows EXE ...
        Exe manifest file generated in C:\Users\fvarrui\GitHub\HelloWorldGradle\build\assets\HelloWorldGradle.exe.manifest!
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\assets\HelloWorldGradle.exe.manifest] to file [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\launch4j\app.exe.manifest]
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\assets\windows\HelloWorldGradle.ico] to file [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\launch4j\app.ico]
        Copying file [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\HelloWorldGradle-1.0.0-runnable.jar] to file [C:\Users\fvarrui\GitHub\HelloWorldGradle\build\launch4j\app.jar]

> Task :packageForWindows FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':packageForWindows'.
> Could not create task ':launch4j_6e89fff5-0d4e-4571-9e5f-016a43679354'.
   > Could not create task of type 'Launch4jLibraryTask'.
      > Cannot run Project.afterEvaluate(Closure) when the project is already evaluated.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 16s
4 actionable tasks: 4 executed
```

Maybe it would be convenient to [report this bug in the original project](https://github.com/TheBoegl/gradle-launch4j/issues).

--- a/src/main/java/io/github/fvarrui/javapackager/gradle/CreateWindowsExe.java
+++ b/src/main/java/io/github/fvarrui/javapackager/gradle/CreateWindowsExe.java
@@ -3,7 +3,6 @@ package io.github.fvarrui.javapackager.gradle;
 import java.io.File;
 import java.util.HashSet;
 import java.util.List;
-import java.util.UUID;
 
 import org.apache.commons.lang3.StringUtils;
 
@@ -53,7 +52,8 @@ public class CreateWindowsExe extends WindowsArtifactGenerator {
 		
 		String jarPath = winConfig.isWrapJar() ? genericJar.getAbsolutePath() : jarFile.getName();
 		
-		Launch4jLibraryTask l4jTask = createLaunch4jTask();
+		Launch4jLibraryTask l4jTask = Context.getGradleContext().getLibraryTask();
+		l4jTask.getOutputs().upToDateWhen(task -> false);
 		l4jTask.setHeaderType(winConfig.getHeaderType().toString());
 		l4jTask.setJar(jarPath);
 		l4jTask.setDontWrapJar(!winConfig.isWrapJar());
@@ -85,10 +85,6 @@ public class CreateWindowsExe extends WindowsArtifactGenerator {
 		FileUtils.copyFileToFile(genericExe, executable);
 		
 		return executable;
-	}
-	
-	private Launch4jLibraryTask createLaunch4jTask() {
-		return Context.getGradleContext().getProject().getTasks().create("launch4j_" + UUID.randomUUID(), Launch4jLibraryTask.class);
 	}
 	
 	private void createAssets(WindowsPackager packager) throws Exception {

--- a/src/main/java/io/github/fvarrui/javapackager/gradle/GradleContext.java
+++ b/src/main/java/io/github/fvarrui/javapackager/gradle/GradleContext.java
@@ -1,7 +1,10 @@
 package io.github.fvarrui.javapackager.gradle;
 
 import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
 
+import edu.sc.seis.launch4j.tasks.Launch4jLibraryTask;
 import org.gradle.api.Project;
 import org.gradle.api.logging.Logger;
 
@@ -14,6 +17,8 @@ import io.github.fvarrui.javapackager.packagers.Packager;
 public class GradleContext extends Context<Logger> {
 
 	private Project project;
+
+	private Launch4jLibraryTask libraryTask;
 	
 	public GradleContext(Project project) {
 		super();
@@ -67,6 +72,14 @@ public class GradleContext extends Context<Logger> {
 	@Override
 	public File createWindowsExe(Packager packager) throws Exception {
 		return new CreateWindowsExe().apply(packager);	
+	}
+
+	public Launch4jLibraryTask getLibraryTask() {
+		return libraryTask;
+	}
+
+	public void setLibraryTask(Launch4jLibraryTask libraryTask) {
+		this.libraryTask = libraryTask;
 	}
 
 }

--- a/src/main/java/io/github/fvarrui/javapackager/gradle/PackagePlugin.java
+++ b/src/main/java/io/github/fvarrui/javapackager/gradle/PackagePlugin.java
@@ -1,9 +1,11 @@
 package io.github.fvarrui.javapackager.gradle;
 
+import edu.sc.seis.launch4j.tasks.Launch4jLibraryTask;
+import io.github.fvarrui.javapackager.packagers.Context;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 
-import io.github.fvarrui.javapackager.packagers.Context;
+import java.util.UUID;
 
 /**
  * JavaPackager Gradle plugin
@@ -24,6 +26,8 @@ public class PackagePlugin implements Plugin<Project> {
 		
 		project.getExtensions().create(SETTINGS_EXT_NAME, PackagePluginExtension.class, project);
 		project.getTasks().create(PACKAGE_TASK_NAME, DefaultPackageTask.class).dependsOn("build");
+
+		Context.getGradleContext().setLibraryTask(project.getTasks().create("launch4j_" + UUID.randomUUID(), Launch4jLibraryTask.class));
 
 	}
 

